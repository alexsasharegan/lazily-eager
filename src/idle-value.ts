import { Option, None, Some } from "safe-types";
import { reqIdleCb, cancelIdleCb } from "./shims/request-idle-callback";
import { identity } from "safe-types/lib/utils";

export {
  reqIdleCb,
  cancelIdleCb,
  CancelIdleCallback,
  IdleDeadline,
  RequestIdleCallback,
} from "./shims/request-idle-callback";

/**
 * Construct an `IdleValue` of type `T` with a function that creates & returns
 * the value, `T`. It will be enqueued to run when the browser is idle, but if
 * accessed before it's created, it will cancel the callback and run the
 * function immediately.
 */
export class IdleValue<T> {
  private option: Option<T> = None();
  private handle: Option<number>;

  constructor(private factory: () => T) {
    this.handle = Some(
      reqIdleCb(() => {
        this.option = Some(this.factory());
      })
    );
  }

  private cancelIdleCb(): void {
    this.handle = this.handle.and_then<number>(handle => {
      cancelIdleCb(handle);
      return None();
    });
  }

  get value(): T {
    return this.option.match({
      Some: identity,
      None: () => {
        this.cancelIdleCb();

        const value = this.factory();
        this.option = Some(value);

        return value;
      },
    });
  }

  set value(new_value: T) {
    this.cancelIdleCb();
    this.option = Some(new_value);
  }
}
