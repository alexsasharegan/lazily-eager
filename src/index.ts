export {
  IdleValue,
  CancelIdleCallback,
  cancelIdleCb,
  IdleDeadline,
  reqIdleCb,
  RequestIdleCallback,
} from "./idle-value";
