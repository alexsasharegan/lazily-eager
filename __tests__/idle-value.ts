import { IdleValue } from "../src/index";

describe("IdleValue", async () => {
  it("should work", async () => {
    let formatter = new IdleValue(
      () =>
        new Intl.DateTimeFormat("en-US", {
          timeZone: "America/Phoenix",
        })
    );

    await new Promise(r => setTimeout(r, 50));

    expect(formatter.value.format(0)).toMatchSnapshot();

    formatter.value = new Intl.DateTimeFormat("en-US", {
      timeZone: "America/Los_Angeles",
    });
    expect(formatter.value.format(0)).toMatchSnapshot();

    formatter = new IdleValue(
      () =>
        new Intl.DateTimeFormat("en-US", {
          timeZone: "America/Phoenix",
        })
    );
    expect(formatter.value.format(0)).toMatchSnapshot();
  });
});
