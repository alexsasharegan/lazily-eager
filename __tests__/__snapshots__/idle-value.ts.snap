// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdleValue should work 1`] = `"12/31/1969"`;

exports[`IdleValue should work 2`] = `"12/31/1969"`;

exports[`IdleValue should work 3`] = `"12/31/1969"`;
